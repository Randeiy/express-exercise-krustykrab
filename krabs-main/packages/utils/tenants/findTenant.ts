import { Tenant, DomainSpec } from '../config/config';
import { safeEnv } from '../env';

export function findTenant(tenants: Tenant[], hostname: string): Tenant | undefined {
  return tenants.find((tenant: Tenant) => {
    const domains = tenant.domains.reduce((acc, current) => {
      const currentEnvDomains = current?.[safeEnv] ?? {};
      return [...acc, currentEnvDomains] as never;
    }, []) as DomainSpec[];

    if (domains.includes(hostname)) {
      return true;
    } else {
      const regexDomains = domains.filter(
        (domain: DomainSpec) => domain instanceof RegExp,
      ) as RegExp[];
      const match = regexDomains.map((domain: RegExp) => domain.test(hostname)).filter(Boolean);
      return match.length;
    }
  });
}

export default findTenant;
