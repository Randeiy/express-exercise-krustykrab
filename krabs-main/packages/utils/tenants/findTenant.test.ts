import { findTenant } from './findTenant';

describe('findTenant', () => {
  it('should find the right tenant', () => {
    const config = [
      {
        name: 'website-1',
        domains: [
          {
            dev: 'local.website-1.com',
            test: 'test.website-1.com',
          },
          {
            development: 'local.dev.website-2.com',
            prod: 'prod.website-2.com',
            test: 'test.website-2.com',
          },
        ],
      },
      {
        name: 'website-2',
        domains: [
          {
            dev: /local\.website-2\.[a-z]*\.com/,
            test: /test\.website-2\.[a-z]*\.com/,
          },
        ],
      },
    ];

    expect(findTenant(config, 'test.website-1.com')).toMatchSnapshot();
    expect(findTenant(config, 'test.website-2.example.com')).toMatchSnapshot();
    expect(findTenant(config, 'test.website-2.foo.com')).toMatchSnapshot();
    expect(findTenant(config, 'test.website-2.bar.com')).toMatchSnapshot();
    expect(findTenant(config, 'test.website-2.krabs.com')).toMatchSnapshot();
  });
});

test('Testing Krabs "findTenant" against a large number of vhosts', async () => {
  let config = [];

  for (let i = 0; i < 1_000_000; i++) {
    config.push({
      name: `website-${i}.com`,
      domains: [
        {
          test: `test.${i}.website.com`,
          dev: `local.${i}.website.com`,
        },
        {
          test: `test.${i}.${i}.website.com`,
        },
      ],
    });
  }

  expect(await findTenant(config, 'test.1.website.com')).toMatchSnapshot();
  expect(await findTenant(config, 'test.999.999.website.com')).toMatchSnapshot();
  expect(await findTenant(config, 'test.99999.website.com')).toMatchSnapshot();
  expect(await findTenant(config, 'test.9999999.website.com')).toMatchSnapshot();
  expect(await findTenant(config, 'test.9999999.9999.website.com')).toMatchSnapshot();
});
