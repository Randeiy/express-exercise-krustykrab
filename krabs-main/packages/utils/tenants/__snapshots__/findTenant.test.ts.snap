// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Krabs "findTenant" against a large number of vhosts 1`] = `
Object {
  "domains": Array [
    Object {
      "dev": "local.1.website.com",
      "test": "test.1.website.com",
    },
    Object {
      "test": "test.1.1.website.com",
    },
  ],
  "name": "website-1.com",
}
`;

exports[`Testing Krabs "findTenant" against a large number of vhosts 2`] = `
Object {
  "domains": Array [
    Object {
      "dev": "local.999.website.com",
      "test": "test.999.website.com",
    },
    Object {
      "test": "test.999.999.website.com",
    },
  ],
  "name": "website-999.com",
}
`;

exports[`Testing Krabs "findTenant" against a large number of vhosts 3`] = `
Object {
  "domains": Array [
    Object {
      "dev": "local.99999.website.com",
      "test": "test.99999.website.com",
    },
    Object {
      "test": "test.99999.99999.website.com",
    },
  ],
  "name": "website-99999.com",
}
`;

exports[`Testing Krabs "findTenant" against a large number of vhosts 4`] = `undefined`;

exports[`Testing Krabs "findTenant" against a large number of vhosts 5`] = `undefined`;

exports[`findTenant should find the right tenant 1`] = `
Object {
  "domains": Array [
    Object {
      "dev": "local.website-1.com",
      "test": "test.website-1.com",
    },
    Object {
      "development": "local.dev.website-2.com",
      "prod": "prod.website-2.com",
      "test": "test.website-2.com",
    },
  ],
  "name": "website-1",
}
`;

exports[`findTenant should find the right tenant 2`] = `
Object {
  "domains": Array [
    Object {
      "dev": /local\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
      "test": /test\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
    },
  ],
  "name": "website-2",
}
`;

exports[`findTenant should find the right tenant 3`] = `
Object {
  "domains": Array [
    Object {
      "dev": /local\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
      "test": /test\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
    },
  ],
  "name": "website-2",
}
`;

exports[`findTenant should find the right tenant 4`] = `
Object {
  "domains": Array [
    Object {
      "dev": /local\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
      "test": /test\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
    },
  ],
  "name": "website-2",
}
`;

exports[`findTenant should find the right tenant 5`] = `
Object {
  "domains": Array [
    Object {
      "dev": /local\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
      "test": /test\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
    },
  ],
  "name": "website-2",
}
`;
